File: src/main/java/me/apqx/algorithm/recursion/Recursion.java
package me.apqx.algorithm.recursion;

public class Recursion {
    public static void main(String[] args) {
        String str = "12345";
        System.out.println("num = " + str);
        System.out.println("----printRecursion----");
        printStrRecursion(str);
        System.out.println("----printIteration----");
        printStrIteration(str);
    }

    /**
     * 递归的打印字符串的每一个字符
     */
    static void printStrRecursion(String str) {
        if (str.length() == 1) {
            System.out.println(str);
        } else {
            System.out.println(str.charAt(0));
            printStrRecursion(str.substring(1));
        }
    }

    /**
     * 迭代的打印字符串的每一个字符
     */
    static void printStrIteration(String str) {
        for (int i = 0; i < str.length(); i++) {
            System.out.println(str.charAt(i));
        }
    }
}

File: src/main/kotlin/me/apqx/algorithm/recursion/Recursion.kt
package me.apqx.algorithm.recursion

fun main(args: Array<String>) {
    val str = "12345"
    println("num = $str")
    println("----printRecursion----")
    printStrRecursion(str)
    println("----printIteration----")
    printStrIteration(str)
}

/**
 * 递归的打印字符串的每一个字符
 */
fun printStrRecursion(str: String) {
    if (str.length == 1) {
        println(str)
    } else{
        println(str.toCharArray()[0])
        printStrRecursion(str.substring(1))
    }
}

/**
 * 迭代的打印字符串的每一个字符
 */
fun printStrIteration(str: String) {
    for (i in 0 until str.length) {
        println(str.toCharArray()[i])
    }
}